pkg/task/registration.go

```diff
@@ -0,0 +1,134 @@
+	group          *sync.WaitGroup
+	platformClient platform.Client
+	smClient       sm.Client
+	smHost         string
+}
+
+type serviceBrokerReg struct {
+	*platform.ServiceBroker
+	SmID string
+}
+
+func New(group *sync.WaitGroup, platformClient platform.Client, smClient sm.Client, smHost string) *SBProxyRegistration {
+	return &SBProxyRegistration{
+		group:          group,
+		platformClient: platformClient,
+		smClient:       smClient,
+		smHost:         smHost,
+	}
+}
+
+func (r SBProxyRegistration) Run() {
+	r.group.Add(1)
+	defer r.group.Done()
+	r.run()
+}
+
+func (r SBProxyRegistration) run() {
+	logrus.Debug("Running broker registration task...")
+	registeredBrokers, err := r.platformClient.GetBrokers()
+	if err != nil {
+		logrus.Error("An error occurred while obtaining already registered brokers: ", err)
+		return
+	}
+
+	brokersFromPlatform := make([]serviceBrokerReg, len(registeredBrokers))
+	for _, broker := range registeredBrokers {
+		if !r.isBrokerProxy(broker) {
+			logrus.WithFields(logFields(&broker)).Debug("Registration task SKIPPING registered broker as is not recognized to be proxy broker...")
+			continue
+		}
+		brokerReg := serviceBrokerReg{
+			ServiceBroker: &broker,
+			SmID:          broker.BrokerURL[strings.LastIndex(broker.BrokerURL, "/")+1:],
+		}
+		brokersFromPlatform = append(brokersFromPlatform, brokerReg)
+	}
+
+	proxyBrokers, err := r.smClient.GetBrokers()
+	if err != nil {
+		logrus.Error("An error occurred while obtaining brokers that have to be registered at the platform: ", err)
+		return
+	}
+
+	brokersFromSM := make([]serviceBrokerReg, len(proxyBrokers))
+	for _, broker := range proxyBrokers {
+		brokerReg := serviceBrokerReg{
+			ServiceBroker: &broker,
+			SmID:          broker.Guid,
+		}
+		brokersFromSM = append(brokersFromSM, brokerReg)
+	}
+
+	updateBrokerRegistrations(r.createBrokerRegistration, brokersFromSM, brokersFromPlatform)
+	updateBrokerRegistrations(r.deleteBrokerRegistration, brokersFromPlatform, brokersFromSM)
+}
+
+func (r SBProxyRegistration) deleteBrokerRegistration(broker *platform.ServiceBroker) {
+	logrus.WithFields(logFields(broker)).Info("Registration task will attempt to delete broker...")
+
+	deleteRequest := &platform.DeleteServiceBrokerRequest{
+		Guid: broker.Guid,
+	}
+	if err := r.platformClient.DeleteBroker(deleteRequest); err != nil {
+		logrus.WithFields(logFields(broker)).Error("Error during broker deletion: ", err)
+		//TODO how do we recover from that? Maybe atleast send email / slack notification?
+	} else {
+		logrus.WithFields(logFields(broker)).Info("Registration task broker deletion successful")
+
+	}
+}
+
+func (r SBProxyRegistration) createBrokerRegistration(broker *platform.ServiceBroker) {
+	logrus.WithFields(logFields(broker)).Info("Registration task will attempt to create broker...")
+	createRequest := &platform.CreateServiceBrokerRequest{
+		Name:      ProxyBrokerPrefix + broker.Guid,
+		BrokerURL: r.smHost + "/" + broker.Guid,
```

Comment:
So the broker is registered with the SM URL (you should rename `smHost`).
This means the platform will call directly SM and will bypass the proxy. Which is ok as it eliminates one network hop.
But then the _proxy_ does not justify its name. Actually no one calls the proxy. Probably we should rename it it to _agent_ or something similar.
```
platform <- proxy -> SM
```

... wait a moment ... 😕 

Argh! Now I see. Actually `smHost` is `SbproxyConfig().Host`.
https://github.com/Peripli/service-broker-proxy/pull/3/files#diff-2756e40316edd9a9e9fb9ecf44c7e50cR75
This is very misleading!

Reply:
Its a naming typo - the variable name should be proxyHost